File: mlir/circt/frontends/PyCDE/integration_test/esi_ram.py
# REQUIRES: esi-cosim
# RUN: rm -rf %t
# RUN: %PYTHON% %s %t 2>&1
# ... can't glob *.sv because PyCDE always includes driver.sv, but that's not the
# top that we want to use. Just delete it.
# RUN: rm -f %t/hw/driver.sv
# RUN: esi-cosim-runner.py --no-aux-files --tmpdir %t --schema %t/hw/schema.capnp %s %t/hw/*.sv
# PY: from esi_ram import run_cosim
# PY: run_cosim(tmpdir, rpcschemapath, simhostport)

import pycde
from pycde import (Clock, Input, Module, generator, types)
from pycde.constructs import Wire
from pycde import esi
from pycde.bsp import cosim

import sys

RamI64x8 = esi.DeclareRandomAccessMemory(types.i64, 8)
WriteType = RamI64x8.write.to_server_type


@esi.ServiceDecl
class MemComms:
  write = esi.FromServer(WriteType)
  read = esi.ToFromServer(to_server_type=types.i64, to_client_type=types.i3)
  loopback = esi.ToFromServer(to_server_type=WriteType,
                              to_client_type=WriteType)


class Mid(Module):
  clk = Clock(types.i1)
  rst = Input(types.i1)

  @generator
  def construct(ports):
    (address_chan, ready) = RamI64x8.read.to_server_type.wrap(2, True)
    (read_data, read_valid) = RamI64x8.read(address_chan).unwrap(True)
    (write_data, _) = WriteType.wrap({
        'data': read_data,
        'address': 3
    }, read_valid)
    RamI64x8.write(write_data)


class Top(Module):
  clk = Clock(types.i1)
  rst = Input(types.i1)

  @generator
  def construct(ports):
    Mid(clk=ports.clk, rst=ports.rst)
    RamI64x8.write(MemComms.write("write"))
    read_address = Wire(RamI64x8.read.to_server_type)
    read_data = RamI64x8.read(read_address)
    read_address.assign(MemComms.read(read_data, "read"))

    loopback_wire = Wire(WriteType)
    loopback_wire.assign(MemComms.loopback(loopback_wire, "loopback"))

    RamI64x8.instantiate_builtin("sv_mem",
                                 result_types=[],
                                 inputs=[ports.clk, ports.rst])


def run_cosim(tmpdir=".", schema_path="schema.capnp", rpchostport=None):
  import os
  sys.path.append(os.path.join(tmpdir, "runtime"))
  import ESIMem as esi_sys
  from ESIMem.common import Cosim
  if rpchostport is None:
    port = open("cosim.cfg").read().split(':')[1].strip()
    rpchostport = f"localhost:{port}"

  cosim = Cosim(schema_path, rpchostport)
  print(cosim.list())
  top = esi_sys.top(cosim).bsp
  print(dir(top))

  write_cmd = {"address": 2, "data": 42}
  loopback_result = top.loopback[0](write_cmd)
  assert loopback_result == write_cmd

  read_result = top.read[0](2)
  assert read_result == 0
  read_result = top.read[0](3)
  assert read_result == 0

  top.write[0].write(write_cmd)
  read_result = top.read[0](2)
  assert read_result == 42
  read_result = top.read[0](3)
  assert read_result == 42


if __name__ == "__main__":
  s = pycde.System([cosim.CosimBSP(Top)],
                   name="ESIMem",
                   output_directory=sys.argv[1])
  s.compile()
  s.package()

File: mlir/circt/frontends/PyCDE/integration_test/esi_ram_cpp/esi_ram.cpp
// REQUIRES: esi-cosim
// XFAIL: *

// clang-format off

// Create ESI system
// RUN: rm -rf %t
// RUN: %PYTHON% %S/../esi_ram.py %t 2>&1

// Build the project using the CMakeLists.txt from this directory. Just move
// everything to the output folder in the build directory; this is very convenient
// if we want to run the build manually afterwards.
// RUN: cp %s %t
// RUN: cp %S/CMakeLists.txt %t
// RUN: cmake -S %t \
// RUN:   -B %t/build \
// RUN:   -DCIRCT_DIR=%CIRCT_SOURCE% \
// RUN:   -DPYCDE_OUT_DIR=%t
// RUN: cmake --build %t/build

// Run test
// ... can't glob *.sv because PyCDE always includes driver.sv, but that's not the
// top that we want to use. Just delete it.
// RUN: rm %t/hw/driver.sv
// RUN: esi-cosim-runner.py --tmpdir=%t  \
// RUN:     --no-aux-files               \
// RUN:     --schema %t/hw/schema.capnp  \
// RUN:     --exec %t/build/esi_ram_test \
// RUN:     %t/hw/*.sv

// To run this test manually:
// 1. run `ninja check-pycde-integration` (this will create the output folder, run PyCDE, ...)
// 2. navigate to %t
// 3. In a separate terminal, run esi-cosim-runner.py in server only mode:
//   - cd %t
//   - esi-cosim-runner.py --tmpdir=$(pwd) --schema=$(pwd)/hw/schema.capnp --server-only $(pwd)/hw/top.sv $(pwd)
// 4. In another terminal, run the test executable. When running esi-cosim-runner, it'll print the $port which
//    the test executable should connect to.
//   - cd %t/build
//   - ./esi_ram_test localhost:$port ../hw/schema.capn

// clang-format on
#include <iostream>
#include <memory>
#include <string>
#include <vector>

#include "esi/backends/capnp.h"

#include ESI_COSIM_CAPNP_H

#include "ESISystem.h"

using namespace esi;
using namespace runtime;

template <typename T>
int logTestFailure(T expected, T actual, int testID) {
  std::cerr << "Test " << testID << " failed: expected " << expected << ", got "
            << actual << std::endl;
  return testID;
}

template <typename TBackend>
int runTest(TBackend &backend) {
  // Connect the ESI system to the provided backend.
  esi::runtime::top top(backend);

  auto write_cmd =
      ESITypes::Struct16871797234873963366{.address = 2, .data = 42};

  auto loopback_result = (*top.bsp->loopback)(write_cmd);
  if (loopback_result != write_cmd)
    return logTestFailure(write_cmd, loopback_result, 1);

  auto read_result = (*top.bsp->read)(2);
  if (read_result != ESITypes::I64(0))
    return logTestFailure(ESITypes::I64(0), read_result, 2);

  read_result = (*top.bsp->read)(3);
  if (read_result != ESITypes::I64(0))
    return logTestFailure(ESITypes::I64(0), read_result, 3);

  (*top.bsp->write)(write_cmd);
  read_result = (*top.bsp->read)(2);
  if (read_result != ESITypes::I64(42))
    return logTestFailure(ESITypes::I64(42), read_result, 4);

  read_result = (*top.bsp->read)(3);
  if (read_result != ESITypes::I64(42))
    return logTestFailure(ESITypes::I64(42), read_result, 5);

  // Re-write a 0 to the memory (mostly for debugging purposes to allow us to
  // keep the server alive and rerun the test).
  write_cmd = ESITypes::Struct16871797234873963366{.address = 2, .data = 0};
  (*top.bsp->write)(write_cmd);
  read_result = (*top.bsp->read)(2);
  if (read_result != ESITypes::I64(0))
    return logTestFailure(ESITypes::I64(0), read_result, 6);

  return 0;
}

int run_cosim_test(const std::string &host, unsigned port) {
  // Run test with cosimulation backend.
  esi::runtime::cosim::CapnpBackend cosim(host, port);
  return runTest(cosim);
}

int main(int argc, char **argv) {
  std::string rpchostport;
  if (argc != 3) {
    // Schema not currently used but required by the ESI cosim tester
    std::cerr
        << "usage: esi_ram_test {rpc hostname}:{rpc port} {path to schema}"
        << std::endl;
    return 1;
  }

  rpchostport = argv[1];

  // Parse the RPC host and port from the command line.
  auto colon = rpchostport.find(':');
  if (colon == std::string::npos) {
    std::cerr << "Invalid RPC host:port string: " << rpchostport << std::endl;
    return 1;
  }
  auto host = rpchostport.substr(0, colon);
  auto port = stoi(rpchostport.substr(colon + 1));

  auto res = run_cosim_test(host, port);
  if (res != 0) {
    std::cerr << "Test failed with error code " << res << std::endl;
    return 1;
  }
  std::cout << "Test passed" << std::endl;
  return 0;
}
